File: packages/e2e-test/src/__tests__/conflict.test.ts
/**
 * End-to-End Tests for conflict detection and handling.
 *
 * These tests verify that the system correctly:
 * - Detects conflicts with existing non-generator files
 * - Prevents overwriting conflicting files during generate
 * - Reports detailed conflict information to the user
 * - Handles the --overwrite flag
 */
import { beforeAll, describe, expect, it } from "bun:test";

import { Architecture, Platform } from "@dotfiles/core";
import fs from "node:fs";
import path from "node:path";
import { GITHUB_RELEASE_TOOL, withMockServer } from "./helpers/mock-server";
import { TestHarness } from "./helpers/TestHarness";
import type { ITestTarget } from "./helpers/types";

describe("E2E: conflict detection and handling", () => {
  withMockServer((b) => b.withGitHubTool(GITHUB_RELEASE_TOOL));

  const platformConfigs: ReadonlyArray<ITestTarget> = [
    { platform: Platform.MacOS, architecture: Architecture.Arm64, name: "macOS ARM64" },
    { platform: Platform.Linux, architecture: Architecture.X86_64, name: "Linux x86_64" },
  ];

  for (const config of platformConfigs) {
    describe(`${config.name}`, () => {
      const harness: TestHarness = new TestHarness({
        testDir: import.meta.dir,
        configPath: "fixtures/main/config.ts",
        platform: config.platform,
        architecture: config.architecture,
      });

      const conflictingShimPath = path.join(harness.userBinDir, "github-release-tool");

      describe("conflict detection and handling", () => {
        beforeAll(async () => {
          await harness.clean();
        });

        it("should detect conflicts with existing non-generator files", async () => {
          // Create a conflicting file that is NOT a generator shim
          await fs.promises.mkdir(harness.userBinDir, { recursive: true });
          await fs.promises.writeFile(conflictingShimPath, "");

          // Run detect-conflicts command
          const detectResult = await harness.detectConflicts();

          // Should exit with error code 1
          expect(detectResult.code).toBe(1);

          // Should warn about the conflict (output goes to stdout)
          expect(detectResult.stdout).toContain("Conflicts detected with files not owned by the generator");
          expect(detectResult.stdout).toContain("[github-release-tool]");
          expect(detectResult.stdout).toContain(conflictingShimPath);
          expect(detectResult.stdout).toContain("exists but is not a generator shim");
        });

        it("should fail to generate when conflicting file exists", async () => {
          // The conflicting file should still exist from the previous test
          expect(await harness.fileExists(conflictingShimPath)).toBe(true);

          // Run generate command
          const generateResult = await harness.generate();

          // Generate completes successfully but skips the conflicting shim
          expect(generateResult.code).toBe(0);

          // Should error about the conflict
          expect(generateResult.stdout).toContain('Cannot create shim for "github-release-tool"');
          expect(generateResult.stdout).toContain("conflicting file exists at");
          expect(generateResult.stdout).toContain(conflictingShimPath);
          expect(generateResult.stdout).toContain("Use --overwrite to replace it");

          // Shell scripts should still be generated
          await harness.verifyShellScript("zsh");
          await harness.verifyShellScript("bash");
          await harness.verifyShellScript("powershell");

          // The conflicting file should still be the original (not a shim)
          const fileContent = await harness.readFile(conflictingShimPath);
          expect(fileContent).toBe(""); // Empty file we created with touch
        });

        it("should succeed with --overwrite flag when conflicting file exists", async () => {
          // The conflicting file should still exist
          expect(await harness.fileExists(conflictingShimPath)).toBe(true);

          // Run generate with --overwrite flag
          const generateResult = await harness.generate(["--overwrite"]);

          // Should succeed
          expect(generateResult.code).toBe(0);

          // Shim should now exist and be executable
          await harness.verifyShim("github-release-tool");

          // Shell scripts should be generated
          await harness.verifyShellScript("zsh");
          await harness.verifyShellScript("bash");
          await harness.verifyShellScript("powershell");
        });
      });
    });
  }
});

File: tests/e2e/conflict_test.go
package e2e

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func TestE2EConflict(t *testing.T) {
	t.Parallel()

	ms := NewMockServer(t, "main")
	defer ms.Close()

	h := NewTestHarness(t, HarnessOptions{
		ConfigPath: "config.ts",
	})
	h.MockServerURL = ms.Server.URL

	h.CopyFixture("main")

	// Pre-create user-bin folder inside sandbox and put an unmanaged file there
	userBinDir := filepath.Join(h.TempDir, ".generated", "user-bin")
	err := os.MkdirAll(userBinDir, 0755)
	if err != nil {
		t.Fatalf("failed to create user-bin: %v", err)
	}

	conflictingShimPath := filepath.Join(userBinDir, "github-release-tool")
	err = os.WriteFile(conflictingShimPath, []byte("original unmanaged content"), 0644)
	if err != nil {
		t.Fatalf("failed to create conflicting file: %v", err)
	}

	t.Run("should detect conflicts with existing non-generator files", func(t *testing.T) {
		// Run detect-conflicts
		stdout, stderr, exitCode, err := h.RunCommand("detect-conflicts", "--config", h.ConfigPath)
		if err != nil {
			t.Fatalf("detect-conflicts failed: %v", err)
		}
		if exitCode == 0 {
			t.Fatalf("expected detect-conflicts to fail with exit code 1, but got 0")
		}

		output := stdout + stderr
		if !strings.Contains(output, "Conflicts detected with files not owned by the generator") {
			t.Fatalf("expected output to mention conflict detection, but got:\n%s", output)
		}
		if !strings.Contains(output, "github-release-tool") {
			t.Fatalf("expected output to mention github-release-tool, but got:\n%s", output)
		}
		if !strings.Contains(output, "exists but is not a generator shim") {
			t.Fatalf("expected output to contain 'exists but is not a generator shim', but got:\n%s", output)
		}
	})

	t.Run("should fail to generate when conflicting file exists", func(t *testing.T) {
		// Run generate
		stdout, stderr, exitCode, err := h.Generate("-d")
		if err != nil || exitCode != 0 {
			t.Fatalf("generate failed: %v\nstdout: %s\nstderr: %s", err, stdout, stderr)
		}

		output := stdout + stderr
		if !strings.Contains(output, `Cannot create shim for "github-release-tool"`) {
			t.Fatalf("expected output to warn about not creating shim, but got:\n%s", output)
		}
		if !strings.Contains(output, "conflicting file exists at") {
			t.Fatalf("expected output to mention conflicting file, but got:\n%s", output)
		}
		if !strings.Contains(output, "Use --overwrite to replace it") {
			t.Fatalf("expected output to mention --overwrite option, but got:\n%s", output)
		}

		// The conflicting file should remain intact with the original unmanaged content
		content, err := os.ReadFile(conflictingShimPath)
		if err != nil {
			t.Fatalf("failed to read conflicting file: %v", err)
		}
		if string(content) != "original unmanaged content" {
			t.Fatalf("expected conflicting file to keep original content, but got %q", string(content))
		}
	})

	t.Run("should succeed with --overwrite flag when conflicting file exists", func(t *testing.T) {
		// Run generate with --overwrite
		stdout, stderr, exitCode, err := h.Generate("-d", "--overwrite")
		if err != nil || exitCode != 0 {
			t.Fatalf("generate with --overwrite failed: %v\nstdout: %s\nstderr: %s", err, stdout, stderr)
		}

		// Verify the shim was successfully generated and is executable
		h.AssertShimExistsAndExecutable("github-release-tool")
	})
}
